pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,559 @@
+		// Scaling not required.
+		return false, nil
+	}
+
+	if (newIS.Spec.Replicas) > (deployment.Spec.Replicas) {
+		// Scale down.
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newIS, (deployment.Spec.Replicas), deployment)
+		return scaled, err
+	}
+
+	klog.V(3).Infof("reconcile new machine set %s", newIS.Name)
+
+	addedNewReplicasCount := int32(0)
+
+	for _, is := range oldISs {
+		transferredMachineCount := int32(0)
+		// get the machines for the machine set
+		machines, err := dc.machineLister.List(labels.SelectorFromSet(is.Spec.Selector.MatchLabels))
+		if err != nil {
+			return false, err
+		}
+
+		klog.V(3).Infof("machine in old machine set %s: %d", is.Name, len(machines))
+
+		for _, machine := range machines {
+			nodeName, ok := machine.Labels[v1alpha1.NodeLabelKey]
+			if !ok {
+				return addedNewReplicasCount > 0, fmt.Errorf("node label not found for machine %s", machine.Name)
+			}
+
+			node, err := dc.nodeLister.Get(nodeName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					klog.Warningf("Node %s not found for machine %s", nodeName, machine.Name)
+					continue
+				}
+				return addedNewReplicasCount > 0, fmt.Errorf("failed to get node %s for machine %s: %w", nodeName, machine.Name, err)
+			}
+
+			cond := getMachineCondition(machine, v1alpha1.NodeInPlaceUpdate)
+			if cond == nil || cond.Reason != v1alpha1.UpdateSuccessful || node.Labels[v1alpha1.LabelKeyNodeUpdateResult] != v1alpha1.LabelValueNodeUpdateSuccessful {
+				continue
+			}
+
+			klog.V(3).Infof("transferring machine %s to new machine set %s", machine.Name, newIS.Name)
+
+			// removes labels not present in newIS so that the machine is not selected by the old machine set
+			machineNewLabels := MergeStringMaps(MergeWithOverwriteAndFilter(machine.Labels, is.Spec.Selector.MatchLabels, newIS.Spec.Selector.MatchLabels), map[string]string{v1alpha1.LabelKeyNodeUpdateResult: v1alpha1.LabelValueNodeUpdateSuccessful})
+
+			// update the owner reference of the machine to the new machine set and update the labels
+			addControllerPatch := fmt.Sprintf(
+				`{"metadata":{"ownerReferences":[{"apiVersion":"machine.sapcloud.io/v1alpha1","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"labels":{%s},"uid":"%s"}}`,
+				v1alpha1.SchemeGroupVersion.WithKind("MachineSet"),
+				newIS.GetName(), newIS.GetUID(), labelsutil.GetFormatedLabels(machineNewLabels), machine.UID)
+
+			err = dc.machineControl.PatchMachine(ctx, machine.Namespace, machine.Name, []byte(addControllerPatch))
+			if err != nil {
+				klog.V(3).Infof("failed to transfer the ownership of machine %s to new machine set %s", machine.Name, err)
+				// scale up the new machine set to the already added replicas.
+				scaled, _, err2 := dc.scaleMachineSetAndRecordEvent(ctx, newIS, newIS.Spec.Replicas+addedNewReplicasCount, deployment)
+				klog.V(3).Infof("scale up after failure failed %s", err2)
+				if err2 != nil {
+					return addedNewReplicasCount > 0, err2
```

Comment:
when err != nil, the caller does not consider the first boolean return parameter. You can just return `false, err2` here